jupyter_server_proxy/handlers.py

```diff
@@ -255,22 +297,62 @@ async def proxy(self, host, port, proxied_path):
             self.set_status(500)
             self.write(str(response.error))
         else:
-            self.set_status(response.code, response.reason)
+            # Represent the original response as a RewritableResponse object.
+            original_response = RewritableResponse(orig_response=response)
+            
+            # The function (or list of functions) which should be applied to modify the
+            # response.
+            rewrite_response = self.rewrite_response
+
+            # If this is a single function, wrap it in a list.
+            if isinstance(rewrite_response, (list, tuple)):
+                rewrite_responses = rewrite_response
+            else:
+                rewrite_responses = [rewrite_response]
+
+            # To be passed on-demand as args to the rewrite_response functions.
+            optional_args_to_rewrite_function = {
+                'orig_request': req,
```

Comment:
Hi @maresb, regarding @yuvipanda's [comment](https://github.com/jupyterhub/jupyter-server-proxy/pull/301#discussion_r746226931), could `req` here be changed to `self.request`? Knowing the request that the client made would help jupyter-rsession-proxy since it needs to know the client's requested URI as opposed to the one we have inside the proxy request `req`.

Reply:
@maresb As context, there's [an issue](https://github.com/jupyterhub/jupyter-rsession-proxy/issues/97) with RStudio where, in some jupyterhub environments, it redirects people from `/user/so-and-so/rstudio` to `/auth-sign-in` rather than `/user/so-and-so/rstudio/auth-sign-in`. Since `/auth-sign-in` hits jupyterhub and not the single user user, this is a 404. With your rewrite mechanism, jupyter-rsession-proxy can change the Location header, but it needs to know about `/user/so-and-so/rstudio` path, which I think is only in self.request. Let me know if you can think of a better way however.

Reply:
Oh, and per Yuvi's [other comment](https://github.com/jupyterhub/jupyter-server-proxy/pull/301/files#r749024114) maybe it could be `'request': self.request`.